cpp/src/arrow/flight/types.cc

```diff
@@ -463,6 +463,300 @@ arrow::Result<CancelFlightInfoRequest> CancelFlightInfoRequest::Deserialize(
   return out;
 }
 
+// Helper for stringifying maps containing various types
+template <typename T>
+ostream& operator<<(std::map<std::string, T>) {
+  std::stringstream ss;
+
+  ss << '{';
+  std::string sep = "";
+  for (const auto& [k, v] : session_options) {
+    std::cout << sep << '[' << k << "]: '" << v << "', ";  // PHOXME v.ToString() not implemented yet
+    sep = ", ";
+  }
+  ss << '}';
+
+  return ss.str();
+}
+static bool CompareSessionOptionMaps(
```

Comment:
Does this have different semantics than operator== for maps?
https://en.cppreference.com/w/cpp/container/map/operator_cmp
Where does .session_options come from?

Reply:
It's forcing a check of the variant type index, operator== will ignore that and only test arithmetic value equality which is not sufficient here.  We do fall through to using operator== for std::vector<std::string> variant type because that's correct.

Reply:
(I believe I had an extraneous .session_options there from factoring the code out of another location)